tidy3d/components/medium.py

```diff
@@ -82,24 +82,122 @@ def _eps_model(self, frequency: float) -> complex:
 """ Medium Definitions """
 
 
-class NonlinearSpec(ABC, Tidy3dBaseModel):
-    """Abstract specification for adding a nonlinearity to a medium.
+class NonlinearModel(ABC, Tidy3dBaseModel):
+    """Abstract model for a nonlinear material response.
+    Used as part of a :class:`.NonlinearSpec`."""
+
+    def _validate_medium_type(self, medium: AbstractMedium):
+        """Check that the model is compatible with the medium."""
+        if isinstance(medium, AbstractCustomMedium):
+            raise ValidationError(
+                f"'NonlinearModel' of class {type(self)} is not currently supported "
+                f"for medium class {type(medium)}."
+            )
+        if not isinstance(medium, (Medium, DispersiveMedium)):
+            raise ValidationError(
+                f"'NonlinearModel' of class {type(self)} is not currently supported "
+                f"for medium class {type(medium)}."
+            )
+
+    def _validate_medium(self, medium: AbstractMedium):
+        """Any additional validation that depends on the medium"""
+        pass
+
+    def _validate_medium_freqs(self, medium: AbstractMedium, freqs: List[pd.PositiveFloat]) -> None:
+        """Any additional validation that depends on the central frequencies of the sources."""
+        pass
+
+    def _get_n0(
+        self,
+        n0: complex,
+        medium: AbstractMedium,
+        freqs: List[pd.PositiveFloat],
+    ) -> complex:
+        """Get a single value for n0."""
+        # if n0 is not specified, we need to compute it
+        if n0 is None:
+            if len(freqs) == 0:
+                raise SetupError(
+                    f"Class '{self}' cannot determine 'n0' in the absence of sources. "
+                    "Please either specify 'n0' or add sources to the simulation."
```

Comment:
shouldn't this be asking the user to supply frequencies? could be confusing if called directly.

Reply:
I don't think this method will be called directly, since it is private.

Reply:
In the only use case I have in mind, the freqs are coming from the simulation sources